File: python3/summary_ranges.py
"""
Space   : O(1)
Time    : O(n)
"""


class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:
        if not nums:
            return []

        n = len(nums)
        if n == 1:
            return [str(nums[0])]

        ans = []
        s, e = '', ''
        for i in range(n):
            if i == 0:
                s = nums[i]
                continue
            if nums[i] - nums[i-1] == 1:
                e = nums[i]
                continue
            elif nums[i] - nums[i-1] > 1:
                if e == '':
                    ans.append("{}".format(s))
                else:
                    ans.append("{}->{}".format(s, e))
                s = nums[i]
                e = ''

        if e == '':
            ans.append("{}".format(s))
        else:
            ans.append("{}->{}".format(s, e))

        return ans

File: golang/summary_ranges.go
package main

import "fmt"

/*
Space	: O(1)
Time	: O(n)
*/

func summaryRanges(nums []int) []string {
	n := len(nums)
	ans := []string{}

	if n == 0 {
		return ans
	}

	start := -1
	for i := 0; i < n; i++ {
		if start == -1 {
			start = nums[i]
			continue
		}

		if nums[i] != nums[i-1]+1 {
			if nums[i-1] == start {
				ans = append(ans, fmt.Sprintf("%d", start))
			} else {
				ans = append(ans, fmt.Sprintf("%d->%d", start, nums[i-1]))
			}
			start = nums[i]
		}
	}

	if nums[n-1] == start {
		ans = append(ans, fmt.Sprintf("%d", start))
	} else {
		ans = append(ans, fmt.Sprintf("%d->%d", start, nums[n-1]))
	}

	return ans
}
